compiler/rustc_span/src/span_encoding.rs

```diff
@@ -113,33 +113,25 @@ impl Span {
     /// This function must not be used outside the incremental engine.
     #[inline]
     pub fn data_untracked(self) -> SpanData {
-        if self.len_or_tag != LEN_TAG {
-            // Inline format.
-            debug_assert!(self.len_or_tag as u32 <= MAX_LEN);
-            SpanData {
-                lo: BytePos(self.base_or_index),
-                hi: BytePos(self.base_or_index + self.len_or_tag as u32),
-                ctxt: SyntaxContext::from_u32(self.ctxt_or_zero as u32),
-                parent: None,
-            }
-        } else {
-            // Interned format.
-            debug_assert!(self.ctxt_or_zero == 0);
-            let index = self.base_or_index;
-            with_span_interner(|interner| interner.spans[index as usize])
-        }
+        *self.pointer
     }
 }
 
 #[derive(Default)]
 pub struct SpanInterner {
-    spans: FxIndexSet<SpanData>,
+    spans: FxHashSet<&'static SpanData>,
 }
 
 impl SpanInterner {
-    fn intern(&mut self, span_data: &SpanData) -> u32 {
-        let (index, _) = self.spans.insert_full(*span_data);
-        index as u32
+    fn intern(&mut self, span_data: SpanData) -> &'static SpanData {
+        if let Some(span_data) = self.spans.get(&span_data) {
+            span_data
+        } else {
+            let span_data = Box::new(span_data);
+            let span_data = Box::leak(span_data);
```

Comment:
This is a memory leak. Please store references to keep it alive in `SpanInterner` instead.